module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,640 @@
+	// convert job into a block entry
+	block, err := status.JobToBlock(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to block: %w", err))
+	}
+
+	request := &status.BlockEntry{
+		BlockID: block.ID(),
+		Height:  block.Header.Height,
+	}
+
+	err = e.processSealedHeight(ctx, request)
+	if err == nil {
+		complete()
+	}
+
+	// all errors are thrown as irrecoverable errors except context cancellation
+}
+
+// processSealedHeight downloads ExecutionData for the given block height.
+// If the download fails, it will retry forever, using exponential backoff.
+func (e *executionDataRequester) processSealedHeight(ctx irrecoverable.SignalerContext, request *status.BlockEntry) error {
+	backoff, err := retry.NewExponential(e.config.RetryDelay)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to create retry mechanism: %w", err))
+	}
+	backoff = retry.WithCappedDuration(e.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	// the only error returned is ctx.Err()
+	attempt := 0
+	return retry.Do(ctx, backoff, func(context.Context) error {
+		// download execution data for the block
+		err := e.processFetchRequest(ctx, request)
+
+		if attempt > 0 {
+			e.metrics.FetchRetried()
+		}
+		attempt++
+
+		return retry.RetryableError(err)
+	})
+}
+
+func (e *executionDataRequester) processFetchRequest(ctx irrecoverable.SignalerContext, request *status.BlockEntry) error {
+	logger := e.log.With().Str("block_id", request.BlockID.String()).Logger()
+	logger.Debug().Msgf("processing fetch request for block %d", request.Height)
+
+	result, err := e.results.ByBlockID(request.BlockID)
+
+	// By the time the block is sealed, the ExecutionResult must be in the db
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to lookup execution result for block: %w", err))
+	}
+
+	executionData, err := e.fetchExecutionData(ctx, result.ExecutionDataID, request.Height)
+
+	if isInvalidBlobError(err) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data is valid, and not sign the receipt
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("HALTING REQUESTER: invalid execution data found")
+
+		e.status.Halt()
+		e.metrics.Halted()
+		ctx.Throw(ErrRequesterHalted)
+	}
+
+	// Some or all of the blob was missing or corrupt. retry
+	if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		return err
+	}
+
+	// Any other error is unexpected
+	if err != nil {
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("unexpected error fetching execution data")
+
+		ctx.Throw(err)
+	}
+
+	logger.Debug().Msgf("Fetched execution data for block %d", request.Height)
+
+	request.ExecutionData = executionData
```

Comment:
I'm fine with that. In this case, the `ExecutionData` field it recorded as an optimization to avoid a second db lookup when the notifications are sent. If notifications are too far behind the downloads, the `ExecutionData` field is removed to avoid unnecessary memory use.